### ISSUE TITLE
[Bug Report] Some Menu Bugs Encountered

### ISSUE THREAD

**Issue by SakiSakiSakiSakiSaki** [Decimation/SmartImage#39]
Just documenting some bugs from Release 3.0.1 incase you weren't aware.

-  Selecting more than one Search Engine or Priority Engine by mouse, and then hitting "Enter", will deselect all the engines you had selected, except for the last one. 
https://user-images.githubusercontent.com/41088844/222612706-9388d698-4a8e-4e53-bcb8-191cd9c85318.mp4
    - The inverse is also true. Deselecting engines and then hitting "Enter", will reselect them except for the last one. In order to properly deselect them, you have to deselect them one at a time, hitting "Enter" after each deselect.
https://user-images.githubusercontent.com/41088844/222613635-4a8aa95e-8f65-499a-81be-cc1ffaeaf26d.mp4

- Clicking "Auto-search" does not do anything. Saving and hitting "Enter" does not help in this case.
https://user-images.githubusercontent.com/41088844/222613433-45ad280b-f450-41a9-a724-ae042608b4c3.mp4
    - Search through context menu does in fact auto-search, even with the toggle disabled (not an issue).

- If the Input auto-populates from your clipboard, you're unable to search for any other image, even after clicking "Clear" or "Restart". The only solution to this is to copy a random string, and try again from the beginning.
https://user-images.githubusercontent.com/41088844/222614548-7b01f6c9-6c40-4447-8d18-02211a8438b2.mp4

- Sometimes after searching, the entire window will freeze/hang up, and suddenly close. I thought this might've had to do with the file being in the same directory as the program, but same thing occurs when it's the same file in it's own directory.
https://user-images.githubusercontent.com/41088844/222614802-4d02295a-ae3d-4c23-b283-8c1b095b0c6e.mp4
   - This never seemed to happen when I would use the context menu; after a successful search, I just need to click "Delete" once, and the image would be deleted. However _after this_ if I clicked "Clear" and/or "Restart" and drag and dropped a new image and ran that, the freezing issue would occur again.

**Comment by Decimation** [comment]
Thanks for the report! I'll resolve these issues in the next release.

**Comment by Decimation** [comment]
I'm looking into the issues now. Can you reproduce these issues on the previous release? Also, what is your .NET version (you can find out by running `dotnet --info`). Could you also share your SmartImage configuration?

**Comment by SakiSakiSakiSakiSaki** [comment]
Hey friend, sorry for the late response.

I just downloaded version _3.0.0_ for my re-testing (I assume this is the "previous release" you were talking about):


> -  Selecting more than one Search Engine or Priority Engine by mouse, and then hitting "Enter", will deselect all the engines you had selected, except for the last one. 
https://user-images.githubusercontent.com/41088844/222612706-9388d698-4a8e-4e53-bcb8-191cd9c85318.mp4
    - The inverse is also true. Deselecting engines and then hitting "Enter", will reselect them except for the last one. In order to properly deselect them, you have to deselect them one at a time, hitting "Enter" after each deselect.
https://user-images.githubusercontent.com/41088844/222613635-4a8aa95e-8f65-499a-81be-cc1ffaeaf26d.mp4

_These two still occur in v3.0.0._


> - Clicking "Auto-search" does not do anything. Saving and hitting "Enter" does not help in this case.
https://user-images.githubusercontent.com/41088844/222613433-45ad280b-f450-41a9-a724-ae042608b4c3.mp4
    - Search through context menu does in fact auto-search, even with the toggle disabled (not an issue).
    
_Not applicable in v3.0.0._


> - If the Input auto-populates from your clipboard, you're unable to search for any other image, even after clicking "Clear" or "Restart". The only solution to this is to copy a random string, and try again from the beginning.
https://user-images.githubusercontent.com/41088844/222614548-7b01f6c9-6c40-4447-8d18-02211a8438b2.mp4

_This still occurs in v3.0.0. Can't have any path in your clipboard prior to opening the .exe._

> - Sometimes after searching, the entire window will freeze/hang up, and suddenly close. I thought this might've had to do with the file being in the same directory as the program, but same thing occurs when it's the same file in it's own directory.
https://user-images.githubusercontent.com/41088844/222614802-4d02295a-ae3d-4c23-b283-8c1b095b0c6e.mp4 
    - This never seemed to happen when I would use the context menu; after a successful search, I just need to click "Delete" once, and the image would be deleted. However _after this_ if I clicked "Clear" and/or "Restart" and drag and dropped a new image and ran that, the freezing issue would occur again.

_This doesn't seem to occur in v3.0.0. But looking back at the v3.0.1 video I linked to this "freezing" issue, it occurred after I pressed "Delete". Testing v3.0.1 again, I couldn't consistently replicate this issue._


---

> Also, what is your .NET version (you can find out by running dotnet --info)
```
.NET SDK (reflecting any global.json):
 Version:   6.0.101
 Commit:    ef49f6213a
```

---
> Could you also share your SmartImage configuration?

I can't imagine these being of use since they're from fresh downloads due to the testing I just did.
- v.3.0.0
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
    <appSettings>
        <add key="OnTop" value="True" />
        <add key="SearchEngines" value="ImgOps, GoogleImages, TinEye, Iqdb, TraceMoe, KarmaDecay, Yandex, Bing, Tidder, RepostSleuth" />
        <add key="PriorityEngines" value="Auto" />
    </appSettings>
</configuration>
```
- v3.0.1
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
    <appSettings>
        <add key="SearchEngines" value="All" />
        <add key="OnTop" value="True" />
        <add key="PriorityEngines" value="Auto" />
        <add key="OpenRaw" value="False" />
        <add key="EhUsername" />
        <add key="EhPassword" />
    </appSettings>
</configuration>
```

Let me know if there's anything else you need from me!

**Comment by Decimation** [comment]
>>Also, what is your .NET version (you can find out by running dotnet --info)

>.NET SDK (reflecting any global.json):
> Version:   6.0.101
> Commit:    ef49f6213a


SmartImage uses .NET 7. Can you update to .NET 7, then attempt to replicate the issues again?

**Comment by Decimation** [comment]
### Engine configuration

The jankiness of selecting/toggling engine options has always been present AFAIK. This is due to constraints with how the GUI library (`Terminal.Gui`) and mapping a `ListView` to a flagged `enum`. I can rewrite the functionality which does this (it was in the backlog).

### Auto-search

As you correctly identified, auto-search works when using **SmartImage** via the *command-line* and *context menu*, which is mostly what it was intended for. I have been trying to enable its use for the _GUI_ (the main mode), but due to how `Terminal.Gui` works internally, any successful attempts I've made thus far are too hacky and issue-prone, and I don't want to release builds in such states. I'll also try to work on this as well.

### Clipboard

In your video, you drag-and-drop another file into the input field before clearing the input field (use the `[Clear]` button to clear the input field, or `[Restart]` to clean everything up). This results in two inputs (files in this case) in the field:

![2023-04-15_(04h25m33s)_On5r4imlCU](https://user-images.githubusercontent.com/4624891/232205397-80a14e67-d65f-4a2b-869d-2cea56ad2d13.png)

Dragging-and-dropping into text fields appends data, rather than clearing it (`Terminal.Gui` behavior I cannot change).

**SmartImage** should have safeguards in place to ensure the input field contains valid input, but you may have found an edge case I was unaware of. I will see if this is the case. Could you replicate this again, but clear the input field before dragging-and-dropping another file into the input field?

**Comment by SakiSakiSakiSakiSaki** [comment]
> SmartImage uses .NET 7. Can you update to .NET 7, then attempt to replicate the issues again?

Sure, which version do you want me to retest?

**Comment by SakiSakiSakiSakiSaki** [comment]
> In your video, you drag-and-drop another file into the input field before clearing the input field (use the `[Clear]` button to clear the input field, or `[Restart]` to clean everything up). This results in two inputs (files in this case) in the field:

So I just tried this in v3.0.0. Pressing `[Clear]` does clear the input field, but it doesn't populate itself with the new file being dragged in, it remains empty. Hitting `[Run]` does nothing in this state, it says `Error: invalid input` in cyan text. A workaround is to enter and exit `[Config]`, and _then_ drag the file in.

**Comment by Decimation** [comment]
> > In your video, you drag-and-drop another file into the input field before clearing the input field (use the `[Clear]` button to clear the input field, or `[Restart]` to clean everything up). This results in two inputs (files in this case) in the field:
> 
> So I just tried this in v3.0.0. Pressing `[Clear]` does clear the input field, but it doesn't populate itself with the new file being dragged in, it remains empty. Hitting `[Run]` does nothing in this state, it says `Error: invalid input` in cyan text. A workaround is to enter and exit `[Config]`, and _then_ drag the file in. 

This is *technically* the correct behavior, although it is somewhat counterintuitive. 

When you click a button, *focus* is changed to the button (like clicking in and out of windows). You just have to click the input field again (you can tell if it's focused by the cursor underscore `_`). This is also just how the UI library works (as is common in most UI designs). 

However, I have added in a QOL change in the upcoming release which will automatically change focus back to the input field after `[Clear]` is clicked. Similarly, the reason your workaround works is because exiting the `[Config]` dialog automatically focuses the input field again.

**Comment by Decimation** [comment]
> > SmartImage uses .NET 7. Can you update to .NET 7, then attempt to replicate the issues again?
> 
> Sure, which version do you want me to retest?

If you would, try both `3.0.1` and `3.0.0` so I can isolate which problems appear in which version.

**Comment by SakiSakiSakiSakiSaki** [comment]
Apologies for the delayed response,

Results for `v3.0.0` and `v3.0.1` are below after updating to .NET 7:

> -  Selecting more than one Search Engine or Priority Engine by mouse, and then hitting "Enter", will deselect all the engines you had selected, except for the last one. 
https://user-images.githubusercontent.com/41088844/222612706-9388d698-4a8e-4e53-bcb8-191cd9c85318.mp4
    - The inverse is also true. Deselecting engines and then hitting "Enter", will reselect them except for the last one. In order to properly deselect them, you have to deselect them one at a time, hitting "Enter" after each deselect.
https://user-images.githubusercontent.com/41088844/222613635-4a8aa95e-8f65-499a-81be-cc1ffaeaf26d.mp4

_Both occurs in `v3.0.0`._
_Both occurs in `v3.0.1`._

> - Clicking "Auto-search" does not do anything. Saving and hitting "Enter" does not help in this case.
https://user-images.githubusercontent.com/41088844/222613433-45ad280b-f450-41a9-a724-ae042608b4c3.mp4
    - Search through context menu does in fact auto-search, even with the toggle disabled (not an issue).
    
_Not applicable in v3.0.0.._
_Occurs in `v3.0.1`._

> - If the Input auto-populates from your clipboard, you're unable to search for any other image, even after clicking "Clear" or "Restart". The only solution to this is to copy a random string, and try again from the beginning.
https://user-images.githubusercontent.com/41088844/222614548-7b01f6c9-6c40-4447-8d18-02211a8438b2.mp4

_Occurs in `v3.0.0`. Workaround works still. Your suggestion also works._
_Can't get clipboard to populate the input field in `v3.0.1`._

> - Sometimes after searching, the entire window will freeze/hang up, and suddenly close. I thought this might've had to do with the file being in the same directory as the program, but same thing occurs when it's the same file in it's own directory.
https://user-images.githubusercontent.com/41088844/222614802-4d02295a-ae3d-4c23-b283-8c1b095b0c6e.mp4 
    - This never seemed to happen when I would use the context menu; after a successful search, I just need to click "Delete" once, and the image would be deleted. However _after this_ if I clicked "Clear" and/or "Restart" and drag and dropped a new image and ran that, the freezing issue would occur again.

_I've deduced that this is a `v3.0.1` issue, and largely seems to occur after successfully searching an image, and clicking "Delete". The entire window will freeze, and automatically close. Doesn't matter if I "Clear" or "Restart" before clicking "Delete" or not, the same freeze, close, happens._

---

I hope these results find you well!

**Comment by Decimation** [comment]
Please try the [`v3.0.2-pre`](https://github.com/Decimation/SmartImage/releases/tag/v3.0.2-pre) prerelease.

### Changes
- Improves the engine selection UX
- Fixes delete button crash
- UI/UX improvements
- *Ascii2D* seems to be working again (not related to any changes I made; *Ascii2D* must have made server-sided changes which allows **SmartImage** to use it again)

### Notes
- *Auto-search* is still not yet fully operational. I am still working on it.
- The `[Queue]` button and its functionality are not yet implemented -- it is just a placeholder for now
- Other internal changes not yet documented

**Comment by SakiSakiSakiSakiSaki** [comment]
> ### Changes
> * Improves the engine selection UX

So I'm not sure if this addressed the buttons being selected/deselected en masse with pressing `Enter`, but that issue still exists in v.3.0.2
> * Fixes delete button crash

I can't test this out as SmartImage crashes after searching an image:
https://user-images.githubusercontent.com/41088844/233818755-ac4d0689-bb0b-4fb5-b8b9-64b12dde1a66.mp4

The image in question:
![Background 26_waifu2x_art_noise3_scale_tta_1](https://user-images.githubusercontent.com/41088844/233818772-a3431b85-d943-4665-9cab-e58f45cf15e0.png)

---

Not sure if these issues are cause of my specific machine, or if it's just cause you haven't gotten around to them yet.
Thanks for the update!

**Comment by Decimation** [comment]
Try the updated `v3.0.2-pre` build. I just made some changes/fixes:

I forgot to document that you had to check `None` to clear the selections, but I added a button to clear the engine selection. I apologize for the un-intuitive UX in that regard.

https://user-images.githubusercontent.com/4624891/233822278-1d419041-a3ee-4ef3-b78d-95942ee0b82b.mp4

I tried the image you used and it works in the latest build:

https://user-images.githubusercontent.com/4624891/233822365-28986505-6ab9-49e3-a6de-ca5d19d31537.mp4

Let me know!

**Comment by SakiSakiSakiSakiSaki** [comment]
> I forgot to document that you had to check `None` to clear the selections

By `None` do you mean `Clear`?

> I apologize for the un-intuitive UX in that regard.

No worries! I imagine trying to make a UI out of a terminal is going to be frustrating and hacky from the very start. Whatever you're unable to address can be chalked up to library limitations really.

Overall the menu selection is quite solid, with that pressing "Enter" after mass select bug being mostly or entirely gone. I assume you can't stomp it out for deselect, which is probably library limitations again. Not sure if you're aware of this, but selecting multiple engines, and hitting the `Save` button doesn't actually save them. You must press "Enter" to lock your choices in.

> I tried the image you used and it works in the latest build:

So it does work for me as well, but it takes much longer for me than it does for you in your video. Almost 30 seconds after clicking "Run" does the process appear, and then open a window to the result. Wonder if it's cause my CPU was occupied?

**Comment by Decimation** [comment]
> So it does work for me as well, but it takes much longer for me than it does for you in your video. Almost 30 seconds after clicking "Run" does the process appear, and then open a window to the result. Wonder if it's cause my CPU was occupied?

Can you record this issue? When you do so, make sure the entire shell UI is visible (with the status bar on the right).

![VsDebugConsole_2023-04-28_(12h05m53s)_tR0bcpB4Qw](https://user-images.githubusercontent.com/4624891/235210039-0d682ec3-f400-4521-9635-92f8388cb0d9.png)

Also, is this latency a new issue with `v3.0.2-pre`?

**Comment by SakiSakiSakiSakiSaki** [comment]
`v3.0.0` 
https://user-images.githubusercontent.com/41088844/235211882-3d2beaba-31cf-49c2-9dee-a09008dd8310.mp4

`v3.0.1` 
https://user-images.githubusercontent.com/41088844/235212643-3fcc78a5-9596-4102-b9ae-8dd1f16ee4e0.mp4

`v3.0.2` 
Ran this a few times:
First attempt: https://user-images.githubusercontent.com/41088844/235212779-c989dd43-ce01-4473-99aa-d296b131a211.mp4
Second attempt: https://user-images.githubusercontent.com/41088844/235212921-7ebc0dcb-524b-4234-b18a-a43456fac3f0.mp4

I guess it's working fine now. Wonder what was up with last time.

**Comment by Decimation** [comment]
> I guess it's working fine now. Wonder what was up with last time.

It was probably latency with the upload server. Are there any other problems or suggestions you have?

**Comment by SakiSakiSakiSakiSaki** [comment]
> Are there any other problems or suggestions you have?

Covering our bases, 
- Having focus set onto the input field again after pressing [Clear] was something you had planned.
- Queue is something you're working on as well.
- I [mentioned above](https://github.com/Decimation/SmartImage/issues/39#issuecomment-1517155904) that [Delete] would crash the window. Here's another video of that happening:
  - https://user-images.githubusercontent.com/41088844/235217400-351c980a-9118-4afb-8438-573750709b66.mp4
- Auto-search not being togglable has also been mentioned above.
- I mentioned elsewhere, but a toggle to enable/disable the sound effect after a successful search would be great. A single instance is fine, but once you implement the Queue, it would occur once every few seconds.

Other than that, everything else is functioning beautifully.

**Comment by Decimation** [comment]
In regards to the `[Delete]` crash, can you try the updated `v3.0.2-pre` build? I added temporary file logging to diagnose the problem. It should log to `smartimage.log`.

**Comment by SakiSakiSakiSakiSaki** [comment]
Doesn't seem to do anything this time, crash nor delete the file: 
- https://user-images.githubusercontent.com/41088844/235223144-d3783c1f-8a21-45f8-b3f0-d435ac67157e.mp4
- [smartimage.log](https://github.com/Decimation/SmartImage/files/11356256/smartimage.log)

**Comment by Decimation** [comment]
Can you try the updated `v3.0.2-pre` build now? I believe I fixed the `[Delete]` problem, and I implemented *Silent* mode which toggles the taskbar flash and notification sound.

**Comment by SakiSakiSakiSakiSaki** [comment]
`[Run]` appears to be greyed out for me?

https://user-images.githubusercontent.com/41088844/235229141-f9a9a903-645c-48c3-aeda-d5e0ae0a43a0.mp4

**Comment by Decimation** [comment]
> `[Run]` appears to be greyed out for me?
> 2023-04-28.14-48-06.mp4

Sorry, can you try the new build again? I fixed that as well.

**Comment by SakiSakiSakiSakiSaki** [comment]
Works now! Awesome!

**Comment by Decimation** [comment]
I'm closing this as it seems everything has been resolved. Feel free to reopen if there are any bugs you encounter. Could you also describe how you wanted the queue to be implemented in #40? Also, `v3.0.2` has been released.

**Comment by SakiSakiSakiSakiSaki** [comment]
> I'm closing this as it seems everything has been resolved. Feel free to reopen if there are any bugs you encounter. Could you also describe how you wanted the queue to be implemented in #40? Also, `v3.0.2` has been released.

Hey thanks for that. Just wanted to know if managed to get Auto-search working in `v3.0.2`? It doesn't remain checked when I leave the Config menu.

`Clipboard detection` also seems to retoggle itself after you close the program. Not sure if that's intentional or not.

**Comment by Decimation** [comment]
> Hey thanks for that. Just wanted to know if managed to get Auto-search working in `v3.0.2`? It doesn't remain checked when I leave the Config menu.
> 
> `Clipboard detection` also seems to retoggle itself after you close the program. Not sure if that's intentional or not.

I haven't changed `Auto-search` behavior yet, and `Clipboard detection` isn't part of the config file (yet).
